### PR TITLE
Don't give back completions after stray dots

### LINKED ISSUE #3242: Poor Intellisense/completion with rest parameters
Test case:

function foo(.<request completions>

Observe:
a list of global visible symbols are proposed. however the only possible completion would be another dot.

the language service shouldn't propose anything here. It doesn't propose anything for

function foo(a<code completions>

either

### PR DESCRIPTION
**Pull request by DanielRosenwasser**
Fixes #3242.

### CODE REVIEW

File: src/services/services.ts [hunk 1]
```diff
@@ -2902,23 +2902,30 @@ namespace ts {
...
+                        isRightOfDot = true;
+                    }
+                    else if (parent.kind === SyntaxKind.QualifiedName) {
+                        node = (<QualifiedName>contextToken.parent).left;
+                        isRightOfDot = true;
+                    }
+                    else {
+                        // There is nothing that precedes the dot, so this likely just a stray character
+                        // or leading into a '...' token. Just bail out instead.
+                        return undefined;
```
Comment (mhegazy) [inline src/services/services.ts]: consider moving the check to the top of the function along with other completion blocking checks.

File: src/services/services.ts [hunk 2]
```diff
@@ -2902,23 +2902,30 @@ namespace ts {
...
-                    node = (<QualifiedName>contextToken.parent).left;
-                    isRightOfDot = true;
+            if (contextToken) {
+                let { parent, kind } = contextToken;
+                if (kind === SyntaxKind.DotToken) {
+                    if (parent.kind === SyntaxKind.PropertyAccessExpression) {
+                        node = (<PropertyAccessExpression>contextToken.parent).expression;
+                        isRightOfDot = true;
+                    }
+                    else if (parent.kind === SyntaxKind.QualifiedName) {
```
Comment (mhegazy) [inline src/services/services.ts]: also jsx spread expression should be in this list.. somthing like `<div {.../*here*/`

**Comment by DanielRosenwasser** [comment]
All good now @mhegazy?

**Comment by mhegazy** [comment]
:+1:

**Pull request merged**